dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/zorder/OptimizeBeforeWrite.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.sql.zorder
+
+import org.apache.hadoop.hive.ql.ErrorMsg
+import org.apache.spark.SparkException
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.catalog.HiveTableRelation
+import org.apache.spark.sql.catalyst.expressions.{And, AttributeSet, Expression, Or}
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, InsertIntoStatement, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+case class OptimizeBeforeWrite(session: SparkSession) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan match {
+      case o@OptimizeZorderCommand(child) =>
+        var relation: Option[HiveTableRelation] = None
+        var partitionSpec: Map[String, Option[String]] = Map.empty
+        val hadoopConf = session.sessionState.newHadoopConf()
+        val newChild = child.resolveOperatorsUp {
+          case f@Filter(condition, _) if f.resolved =>
```

Comment:
We can make this code more clearer since we only support two pattern here:
```
case f @ Filter(condition, r: HiveTableRelation) =>
  checkRelation()

case r: HiveTableRelation =>
  checkRelation()

```